src/cargo/util/toml/mod.rs

```diff
@@ -1749,18 +2208,74 @@ impl<P: ResolveToPath> TomlDependency<P> {
             }
             .to_dependency(name, cx, kind),
             TomlDependency::Detailed(ref details) => details.to_dependency(name, cx, kind),
+            TomlDependency::Workspace(_) => unreachable!(),
         }
     }
 
     fn is_version_specified(&self) -> bool {
         match self {
             TomlDependency::Detailed(d) => d.version.is_some(),
             TomlDependency::Simple(..) => true,
+            TomlDependency::Workspace(_) => unreachable!(),
+        }
+    }
+
+    fn is_optional(&self) -> bool {
+        match self {
+            TomlDependency::Detailed(d) => d.optional.unwrap_or(false),
+            TomlDependency::Simple(..) => false,
+            TomlDependency::Workspace(_) => unreachable!(),
+        }
+    }
+
+    fn resolve<'a>(
+        self,
+        cargo_features: &Features,
+        label: &str,
+        inherit: impl FnOnce() -> Option<&'a InheritableFields>,
+        get_ws: impl FnOnce(&InheritableFields) -> Option<TomlDependency<P>>,
+    ) -> CargoResult<TomlDependency<P>> {
+        match self {
+            TomlDependency::Detailed(d) => Ok(TomlDependency::Detailed(d)),
+            TomlDependency::Simple(s) => Ok(TomlDependency::Simple(s)),
+            TomlDependency::Workspace(TomlWorkspaceDependency {
+                workspace: true,
+                features,
+                optional,
+            }) => {
+                cargo_features.require(Feature::workspace_inheritance())?;
+
+                let inherit = inherit().context(format!(
+                    "You cannot inherit fields from a parent workspace currently, tried to on `[dependency.{}]`",
+                    label
+                ))?;
+
+                get_ws(inherit).map_or(
+                    Err(anyhow!("error reading `{label}` from workspace root manifest's `[workspace.{label}]`")),
+                    |dep| {
+                        match dep {
+                            TomlDependency::Simple(s) => Ok(TomlDependency::Simple(s)),
+                            TomlDependency::Detailed(d) => {
+                                let mut dep = d.clone();
+                                dep.add_features(features);
+                                dep.update_optional(optional);
+                                Ok(TomlDependency::Detailed(dep))
+                            }
```

Comment:
Are paths resolve at this point?

Reply:
No, paths aren't resolved at all in part 1. This only works here since we can only inherit from a `Cargo.toml` that has both a `[workspace]` and a `[package]`